### ISSUE TITLE
Angular search parameters ignored, converted to normal query params

### ISSUE THREAD

**Issue by mafrost** [prerender/prerender#197]
For example,
req.url = 

    http://mydomain.com/?_escaped_fragment_=/my/angularpath?mysearchparam=mysearchkey&myotherparam=thiswontwork

req.prerender.url = 

    http:/mydomain.com/?myotherparam=thiswontwork#!/my/angularpath?mysearchparam=mysearchkey

This creates huge issues for me. It's pretty easy to work around this with a temporary fix but thought you should know!

**Comment by thoop** [comment]
```
http://mydomain.com/?_escaped_fragment_=/my/angularpath?mysearchparam=mysearchkey&myotherparam=thiswontwork
```
That URL is incorrectly formatted. It should be:

http://mydomain.com/?_escaped_fragment_=/my/angularpath&mysearchparam=mysearchkey&myotherparam=thiswontwork

Is a crawler accessing that URL? Which middleware are you using?

**Comment by mafrost** [comment]
I am using nginx as a proxy for the prerender server.

Angular search parameters need that question mark.
https://docs.angularjs.org/api/ng/service/$location

    http://example.com/#/some/path?foo=bar&baz=xoxo

Am I missing something? 

My nginx conf is as follows (the interesting parts)
    
        set $prerender 0;
        if ($http_user_agent ~* "baiduspider|twitterbot|facebookexternalhit|rogerbot|linkedinbot|embedly|quora link preview|showyoubot|outbrain|pinterest|slackbot|vkShare|W3C_Validator") {
            set $prerender 1;
        }
        if ($args ~ "_escaped_fragment_") {
            set $prerender 1;
        }
        if ($http_user_agent ~ "Prerender") {
            set $prerender 0;
        }

        if ($uri ~ "\.(js|css|xml|less|png|jpg|jpeg|gif|pdf|doc|txt|ico|rss|zip|mp3|rar|exe|wmv|doc|avi|ppt|mpg|mpeg|tif|wav|mov|psd|ai|xls|mp4|m4a|swf|dat|dmg|iso|flv|m4v|torrent|ttf|woff)") {
            set $prerender 0;
        }
        if ($prerender = 1) {
           rewrite .* /$scheme://$host$request_uri? break;
           proxy_pass http://$host/prerender/http://$host$request_uri;
        }

And with the location block

    location /prerender/ {
        access_log off;
        proxy_cache prerender;
        proxy_cache_valid 200 1440m;
        proxy_cache_valid 404 60m;
        proxy_pass http://localhost:3000/;
        proxy_ignore_headers X-Accel-Expires Expires Cache-Control;
        proxy_ignore_headers Set-Cookie;
        proxy_hide_header Set-Cookie;
        proxy_hide_header X-powered-by;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header Host $http_host;
        expires 1440m;
        proxy_http_version 1.1;
    }

**Comment by thoop** [comment]
Make sure you use the resolver portion of this nginx config to prevent nginx from caching IPs of servers:

https://gist.github.com/thoop/8165802

Also, make sure you are using the hashbang like:
```
http://example.com/#!/some/path?foo=bar&baz=xoxo
```
And then when Google crawls the page, they will transform that URL into:
```
http://example.com/?_escaped_fragment_=/some/path&foo=bar&baz=xoxo
```

Which the Prerender server will then transform back into the original URL with the #!

**Comment by mafrost** [comment]
Yes, I do all of that, and everything works perfectly, except that urls with angular search paramaters in them, for example the one I showed in the first message (and this is used alot in angular apps) are not preserved correctly. Their angular search parameters are converted into "regular" query parameters, and thus, angular can't use them. The problem is in the util.js file https://github.com/prerender/prerender/blob/master/lib/util.js
where the url library "incorrectly" splits the url into different pieces.

**Comment by thoop** [comment]
There is no difference between regular query parameters and angular query parameters. They are the same thing. Is this a bug you're experiencing from angular incorrectly parsing the URL?
https://github.com/angular/angular.js/issues/7239
https://github.com/angular/angular.js/issues/5964

Can you show me the URL that the Prerender server is accessing when you go to:

```
http://example.com/?_escaped_fragment_=/some/path&foo=bar&baz=xoxo
```

Which of these are you seeing accessed?
```
http://example.com/#!/some/path?foo=bar&baz=xoxo
http://example.com/?foo=bar&baz=xoxo#!/some/path
```

**Comment by mafrost** [comment]
I 100% sure it's not the same thing, at least with hashbang-mode.
Try this example if you don't believe me:

    <!DOCTYPE html><html><script src= "http://ajax.googleapis.com/ajax/libs/angularjs/1.3.14/angular.min.js"></script><body ng-app="myapp" ng-controller="appController"><button ng-click="setSearch()">Set search</button><script>

      var app = angular.module('myapp', []);
      app.config(function($locationProvider, $compileProvider) {
          $locationProvider.hashPrefix('!');
        });

      app.controller('appController', function($scope, $location) {
        $scope.setSearch = function() {
          $location.search({myparam1: 1, myparam2: 2});
        }
      });

    </script></body></html>

Save that page and visit it with a query param, for example: 

    http://localhost/myfinefile.html?myqueryparam=hi

And then press that button and you'll see how angular search parameters works. It will update the url to something like

    http://localhost/myfinefile.html?myqueryparam=hi#!?myparam1=1&myparam2=2

But if this url was going through prerender util.js-code, it would end up as

    http://localhost/myfinefile.html?myqueryparam=hi&myparam2=2#!?myparam1=1

Which really isn't what you would want, and would probably mess up your code.

There should be nothing weird with how angular does this, because everything after the hash shouldn't be parsed as a normal query parameter, right? Therefore the error lies within either prerender.io or the url library (which probably wasn't made with SPA SEO in focus).

My workaround for now is to do a custom parse of req.url and replace req.prerender.url with it.

**Comment by thoop** [comment]
You're correct that they are acting differently in hash bang mode.

If you change your example to use HTML5 mode, the URL correctly parses the original query param, for example:
```
http://localhost/myfinefile.html?myqueryparam=hi
```
And then pressing the button it changes the URL to:
```
http://localhost/myfinefile.html?myparam1=1&myparam2=2
```

I'd say this is a difference in how the hashbang and html5mode works with angular but I understand why it works that way. With hash mode, you can't modify the URL without doing a redirect and moving the original query string back behind the hash.

Having two question marks in the URL is not a valid URL and that's why it parses incorrectly. If you look at the issues that I linked, it looks like they tried to fix that issue but I'm not seeing that it ever got merged back into master.

What URL are you seeing sent as req.url?

**Comment by mafrost** [comment]
I think we can close this. It certainly is an invalid URL, which is kind of annoying since hashbang mode is the recommended way to do frontend rendered stuff. To do more advanced apps that depends on search parameters with hashbang mode is SEO suicide apparently. Facebook and many other crawlers actually strips away the queryparams after the hashbang before even sending the url to our prerender server, which makes it even more difficult.

I had to rewrite the codebase to pushstate instead of hashbang and rewrite some build stuff, backend and server configuration to allow for pushstate. But now it works. Thanks for the feedback though!

To sum it up: don't use hashbang mode if your site needs to be crawlable AND has search parameters (#!/myroute/?category=my-category&text=whatever&thiswontwork=true)

**Comment by thoop** [comment]
Glad to hear it works! Yeah, pushstate is the way to go. I'd definitely recommend it over hashbang. Thanks for the update.
